stdlib/Profile/src/Profile.jl

```diff
@@ -349,19 +349,33 @@ function getdict(data::Vector{UInt})
 end
 
 function getdict!(dict::LineInfoDict, data::Vector{UInt})
-    for ip in data
-        # Lookup is expensive, so do it only once per ip.
-        haskey(dict, UInt64(ip)) && continue
-        st = lookup(convert(Ptr{Cvoid}, ip))
-        # To correct line numbers for moving code, put it in the form expected by
-        # Base.update_stackframes_callback[]
-        stn = map(x->(x, 1), st)
-        try Base.invokelatest(Base.update_stackframes_callback[], stn) catch end
-        dict[UInt64(ip)] = map(first, stn)
+    # we don't want metadata here as we're just looking up ips
+    unique_ips = unique(has_meta(data) ? strip_meta(data) : data)
+    n_unique_ips = length(unique_ips)
+    n_unique_ips == 0 && return dict
+    chnl = Channel{Pair{UInt64,Vector{StackFrame}}}(n_unique_ips) do ch
+        # don't use `Threads.@threads` here because we don't want this to get blocked
+        @sync for ips_part in Iterators.partition(unique_ips, div(n_unique_ips, Threads.nthreads(), RoundUp))
+            Threads.@spawn begin
+                for ip in ips_part
+                    put!(ch, UInt64(ip) => _lookup_corrected(ip))
+                end
+            end
+        end
     end
+    foreach(v -> dict[first(v)] = last(v), chnl)
     return dict
 end
 
+function _lookup_corrected(ip::UInt)
+    st = lookup(convert(Ptr{Cvoid}, ip))
+    # To correct line numbers for moving code, put it in the form expected by
+    # Base.update_stackframes_callback[]
+    stn = map(x->(x, 1), st)
+    try Base.invokelatest(Base.update_stackframes_callback[], stn) catch end
+    return map(first, stn)
+end
```

Comment:
A better approach may be to load the callback outside the parallel loop and wrap the whole parallel loop in a `Base.invokelatest() do ... end` block. We can eliminate the dynamic dispatch this way (although the cost may be negligible).

Also, a nitpick: I think #43805 should have waited for @timholy's approval. #43805 is a breaking change for whoever sets `Base.update_stackframes_callback[]` since the callback needs to be race-free after #43805. (Of course, if you guys communicated outside GitHub issue, my nitpick is a meaningless complaint.)

Reply:
No you're right, I merged too eagerly. My bad.

Thanks for the tips

Reply:
btw, just to be clear, #43805 has been reverted

Reply:
Thanks for the clarification. I don't think we really need to be super defensive about `Base.update_stackframes_callback` though. The only important thing is that @timholy knows about this change because, IIUC, this callback is specifically arranged for Revise.

Reply:
Ok. I've removed the channel and simplified based on @timholy's feedback in the other thread.

Thanks, both!